### PR TITLE
[SPARK-27474][CORE] avoid retrying a task failed with CommitDeniedException many times

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

https://issues.apache.org/jira/browse/SPARK-25250 reports a bug that, a task which is failed with `CommitDeniedException` gets retried many times.

This can happen when a stage has 2 task set managers, one is zombie, one is active. A task from the zombie TSM completes, and commits to a central coordinator(assuming it's a file writing task). Then the corresponding task from the active TSM will fail with `CommitDeniedException`. `CommitDeniedException.countTowardsTaskFailures` is false, so the active TSM will keep retrying this task, until the job finishes. This wastes resource a lot.

#21131 firstly implements that a previous successful completed task from zombie `TaskSetManager` could mark the task of the same partition completed in the active `TaskSetManager`. Later #23871 improves the implementation to cover a corner case that, an active `TaskSetManager` hasn't been created when a previous task succeed.

However, #23871 has a bug and was reverted in #24359. With hindsight, #23781 is fragile because we need to sync the states between `DAGScheduler` and `TaskScheduler`, about which partitions are completed.

This PR proposes a new fix:
1. When `DAGScheduler` gets a task success event from an earlier attempt, notify the `TaskSchedulerImpl` about it
2. When `TaskSchedulerImpl` knows a partition is already completed, ask the active `TaskSetManager` to mark the corresponding task as finished, if the task is not finished yet.

This fix covers the corner case, because:
1. If `DAGScheduler` gets the task completion event from zombie TSM before submitting the new stage attempt, then `DAGScheduler` knows that this partition is completed, and it will exclude this partition when creating task set for the new stage attempt. See `DAGScheduler.submitMissingTasks`
2. If `DAGScheduler` gets the task completion event from zombie TSM after submitting the new stage attempt, then the active TSM is already created.

Compared to the previous fix, the message loop becomes longer, so it's likely that, the active task set manager has already retried the task multiple times. But this failure window won't be too big, and we want to avoid the worse case that retries the task many times until the job finishes. So this solution is acceptable.

## How was this patch tested?

a new test case.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @Ngone51 @pgandhi999 @squito @jiangxb1987 @zsxwing 

I think this is a more robust fix.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (pgandhi999) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Doesn't this method need to acquire synchronization on `TaskSchedulerImpl` object?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: no it doesn't. It's called in the thread `task-result-getter`

Comment (pgandhi999) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Correct me if I am wrong, but if I follow the call chain correctly, you are calling the method `markPartitionCompleted` from `enqueuePartitionCompletionNotification`, which is called from `notifyPartitionCompletion`, and which gets invoked inside the DAG Scheduler `event loop thread`. Am I missing something here?

Comment (pgandhi999) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Ahh I see it now, thanks.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (squito) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: taskResultGetter is a multithreaded pool (default 4) so I think you still need extra protection here

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I'm undecided about this part of the change in general (I see an argument for doing it both ways) but I'd still prefer it is not in this PR.

**Review by squito — COMMENTED** [review]
Have you considered making the set of completed partitions a structure that is safe to access from multiple threads, eg. an AtomicBitSet?  the more I think about it, the more that seems like the best solution to deal with how important that set is to multiple threads.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: then how about we use `0` as the task duration here? There is no `taskInfo` so I have to update this part.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: ah good catch! I should do synchronization here.

**Review by Ngone51 — COMMENTED** [review]
This solution just looks like a variant of #21131 to me. And for "best-effort", I think #21131 does better than this fix, since #21131 does not even go through message event loop.

**Comment by cloud-fan** [comment]
Yes it's kind of a variant of #21131. This PR covers the corner case that, an active `TaskSetManager` hasn't been created when a previous task succeed.

There is a corner case that is un-fixable: when a task from a zombie TSM completes, the active TSM has already submitted the task for the same partition. #21131 doesn't cover it, and this PR doesn't cover it either. But this PR does make it more likely to happen, because we go through the event loop, so the active TSM takes longer to know that a partition has completed.

Maybe we can combine the solution? e.g. when a task from zombie TSM completes, notify the active TSM immediately. Later the `DAGScheduler` notify the active TSM again, in case it was not created when the task from zombie TSM completes.

**Comment by Ngone51** [comment]
> The issue cannot be 100% eliminated. Let's say task set 1.0 (zombie) has a running task for a partition, and task set 1.1 (active) has already submitted the task for the same partition and completed. Then there is nothing we can do.

Though we have nothing to do with finding `1`, but what's wrong with it ?

task in TSM 1.1(active) submitted and completed successfully, and TSM 1.0(zombie) has a running task for the same partition. So, task in TSM 1.0 (zombie) would fail later, but it do not have any impact on TSM 1.1(active), and TSM 1.0/1.1 would both finished normally.

**Comment by cloud-fan** [comment]
I think we are discussing the optimization(saving resource) instead of bug? Nothing will go wrong even without #21131

**Comment by cloud-fan** [comment]
> So, task in TSM 1.0 (zombie) would fail later

To correct, I don't think it will fail. It just gets ignored.

**Comment by cloud-fan** [comment]
retest this please

**Comment by Ngone51** [comment]
> Yes it's kind of a variant of #21131. This PR covers the corner case that, an active TaskSetManager hasn't been created when a previous task succeed.

Does this pr really cover the corner case ? Do I miss something ?

> That said, this PR tries to avoid the worst case described above. Even if we go through the event loop now, I don't think it will take a very long time that the task from the active TSM have already re-tried 3 times.

Yeah, I think this pr could avoid the worst case event if we go through the event loop. But, doesn't #21131 could do it either ? What's the advantage compares to #21131 ?

> Maybe we can combine the solution? e.g. when a task from zombie TSM completes, notify the active TSM immediately. Later the DAGScheduler notify the active TSM again, in case it was not created when the task from zombie TSM completed.

I'm neutral on combining the solution since we're doing optimization(saving resource) instead of bug. Though, I'm wodering whether it would do a good help.

**Comment by cloud-fan** [comment]
> But, doesn't #21131 could do it either ? What's the advantage compares to #21131 ?

#21131 did not cover the corner case, that's why we open #23871 . Do I miss something? Maybe you were asking why I prefer this PR over #23871 . The reason is, we don't need to sync the states between `DAGScheduler` and `TaskSchedulerImpl`, which is fragile to me.

**Comment by cloud-fan** [comment]
I've updated the PR description, to make it clearer about what the problem is and why propose this fix.

**Comment by Ngone51** [comment]
No, I'm asking "does this pr really covers the corner case that *an active TaskSetManager hasn't been created when a previous task succeed*." If does, how ? Because I didn't see it from the code.

**Comment by cloud-fan** [comment]
I've added it to the PR description. let me quote it here

> This fix covers the corner case, because:
1. If `DAGScheduler` gets the task completion event from zombie TSM before submitting the new stage attempt, then `DAGScheduler` knows that this partition is completed, and it will exclude this partition when creating task set for the new stage attempt. See `DAGScheduler.submitMissingTasks`
2. If `DAGScheduler` gets the task completion event from zombie TSM after submitting the new stage attempt, then the active TSM is already created.

**Comment by Ngone51** [comment]
Oh, yes, I get to know how this pr could fix the corner case and work finally. It's really a good solution. And now, I would suggest not combine #21131.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala
Comment (pgandhi999) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Just fyi, after adding the synchronized statement, this PR is almost  similar to PR #22806 before it got restructured. I could have simply added synchronized to `markPartitionCompletedInAllTaskSets()` instead of restructuring the PR but did not do so because of the following suggestion by @squito which I agree with.

> I'm not sure how to fix this. You could make TaskSchedulerImpl.markPartitionCompletedInAllTaskSets() synchronized, but then you're getting a lock on the taskSchedulerImpl in the DAGScheduler event loop. That's not good for scheduling throughput, and also want to make sure there is no change of deadlock.

So I would like to once again pose the same question as above: is the scheduling throughput getting impacted by adding the extra synchronization?

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Ah thanks for pointing it out! Yes it's very similar, except that this PR sends message to the task result getter thread pool first, to avoid locking in the DAGScheduler event loop thread.

I don't think we will hit a deadlock. The task result getter thread pool calls `TaskSchedulerImpl.handleSuccessfulTask`, which is synchronized too. And `handleSuccessfulTask` calls method of `TaskSetManager`, which is the same as the newly added `handlePartitionCompleted`.

**Comment by pgandhi999** [comment]
@cloud-fan I tested the PR a couple of times, tests look good. LGTM.

**Comment by Ngone51** [comment]
LGTM

**Comment by jiangxb1987** [comment]
So currently there are three places that tracks the partition status:
1. The MapStatusTracker keeps track of all shuffle partitions, it get updated by DAGScheduler.handleTaskCompletion(), and on FetchFailed it remove the corresponding shuffle partitions;
2. The ShuffleMapStatus keeps track of `pendingPartitions` itself, it's actually a fork of MapStatusTracker, and get cleared on stage submitted;
3. The TaskSetManager keeps track of pending and running tasks, when the number of successful tasks reaches the target number then it mark the TSM (eg. a stage attempt) as successful.

It shall be ideal that we keep track of the pending partitions of each stage in a data structure and update it in a synchronous way. The major problem here is that the DAGScheduler rely on MapStatusTracker to read the shuffle partitions statuses, which is updated asynchronously.

If we don't want to make major change to current infrastructure, the best approach I can think of is to just let DAGScheduler make all the final decision whether a stage has been completed, and all TSM shall update their own status according to that. The only shortcoming here is just there is a time window that some TSM(s) has finished all the tasks but the MapStatusTracker is not yet updated, in this case we shall see unnecessary tasks still running. To further avoid this case we can implement another approach that Wenchen suggested -- To have a status cache for TSM, when a task from zombie TSM completes, notify the active TSM immediately.

**Comment by jiangxb1987** [comment]
The PR looks good from my side. The only concern I have is whether we shall backport this to 2.4? I know the previous PRs are backported, but my feeling is, strictly speaking the PR is a improvement instead of a bugfix, and it's making major change to a relatively critical and fragile part of Spark. Considering there has been one instance that a PR has caused regression of 2.4.1,  personally I would suggest we be more conservative and only include this in master branch. Anyway it's just my two cents, feel free to correct me @squito @cloud-fan

**Review by squito — COMMENTED** [review]
some minor comments but overall I think this is fine.

As for what branch to put it in ... I would have liked to get this into 2.4 (I do think this is a bug fix).  But I'm also pretty embarrassed about the prior regression, and it probably makes more sense to tread carefully here and only put this in master.

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: I think now that you're using the duration across stage attempts for speculation, you don't need to change this test anymore -- but I also don't feel strongly that we really need to keep the old test either.

**Review by squito — APPROVED** [review]
lgtm assuming tests pass

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala
Comment (cloud-fan) [reply core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: This test case creates 2 TSM, then complete the task of one TSM, and expect it to mark the corresponding task of the other TSM to completed.

This is not true anymore, as we need to go through DAGScheduler to let it happen, but this test suite uses a fake DAGScheduler. To simplify the code, I just create one TSM, and then call `markPartitionCompleted` directly.

**Comment by dongjoon-hyun** [comment]
cc @dbtsai

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**
